velox/exec/WindowPartition.cpp

```diff
@@ -136,17 +176,59 @@ std::pair<vector_size_t, vector_size_t> WindowPartition::computePeerBuffers(
     vector_size_t prevPeerStart,
     vector_size_t prevPeerEnd,
     vector_size_t* rawPeerStarts,
-    vector_size_t* rawPeerEnds) const {
-  auto peerCompare = [&](const char* lhs, const char* rhs) -> bool {
-    return compareRowsWithSortKeys(lhs, rhs);
+    vector_size_t* rawPeerEnds) {
+  auto peerCompare =
+      [&](const char* lhs, const char* rhs, RowContainer* data) -> bool {
+    return compareRowsWithSortKeys(lhs, rhs, data);
   };
 
-  VELOX_CHECK_LE(end, numRows());
+  VELOX_CHECK_LE(end, numRows() + startRow());
 
-  auto lastPartitionRow = numRows() - 1;
+  auto lastPartitionRow = numRows() + startRow() - 1;
   auto peerStart = prevPeerStart;
   auto peerEnd = prevPeerEnd;
-  for (auto i = start, j = 0; i < end; i++, j++) {
+
+  auto nextStart = start;
+  if (partial_ && previousGroupLastRow_) {
+    // rowContainer is used to compare whether the current first row is in same
+    // peer with the last row of previous batch.
+    auto rowContainer =
+        std::make_unique<RowContainer>(data_->keyTypes(), data_->pool());
+    auto serialized = BaseVector::create<FlatVector<StringView>>(
+        VARBINARY(), 1, data_->pool());
+    data_->extractSerializedRows(
+        folly::Range(rows_.data() + (start - startRow()), 1), serialized);
+    rowContainer->clear();
+
+    auto firstRow = rowContainer->newRow();
+    rowContainer->storeSerializedRow(*serialized, 0, firstRow);
+    auto secondRow = rowContainer->newRow();
+    rowContainer->storeSerializedRow(*previousGroupLastRow_, 0, secondRow);
+
+    auto peerGroup = peerCompare(secondRow, firstRow, rowContainer.get());
+
+    if (!peerGroup) {
+      peerEnd++;
```

Comment:
This logic is very similar to that in the next for loop. Please can you abstract a common function.

Reply:
@aditi-pandit Yes. Add `findPeerGroupEndIndex `method.